src/lazy/encoder/write_as_ion.rs

```diff
@@ -45,229 +41,210 @@ impl<T> WriteAsIon for T
         writer.write_blob(self)
     }
 }
 
 impl WriteAsIonValue for Clob {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         writer.write_clob(self)
     }
 }
 
-impl<T> WriteAsIonValue for &T
-where
-    T: WriteAsIonValue,
-{
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
+impl<T: WriteAsIonValue> WriteAsIonValue for &T {
+    fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
         (*self).write_as_ion_value(writer)
     }
 }
 
-impl<T> WriteAsIonValue for &[T]
-where
-    for<'x> &'x T: WriteAsIon,
-{
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
-        let mut list = writer.list_writer()?;
-        for value in *self {
-            list.write(value)?;
+macro_rules! impl_write_as_ion_value_for_iterable {
+    ($iterable:ty, $item:ident $(, const $n:ident: $n_type:ty)?) => {
+        impl<$item $(, const $n: $n_type)?> WriteAsIonValue for $iterable
+        where
+            for<'a> &'a $item: WriteAsIon,
+        {
+            fn write_as_ion_value<V: ValueWriter>(&self, writer: V) -> IonResult<()> {
+                writer.write_list(|list| {
+                    for value in self.iter() {
+                        list.write(value)?;
+                    }
+                    Ok(())
+                })
+            }
         }
-        list.end()?;
-        Ok(())
-    }
+    };
 }
 
-impl<T, const N: usize> WriteAsIonValue for [T; N]
+impl_write_as_ion_value_for_iterable!(Vec<T>, T);
+impl_write_as_ion_value_for_iterable!(&[T], T);
+impl_write_as_ion_value_for_iterable!([T; N], T, const N: usize);
+
+pub trait WriteAsSExp<T>: Sized
 where
-    T: WriteAsIonValue,
+    T: WriteAsIon,
 {
-    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
-        &self,
-        writer: V,
-    ) -> IonResult<()> {
-        let mut list = writer.list_writer()?;
-        for value in self {
-            list.write(value)?;
+    fn as_sexp(self) -> SExpTypeHint<Self, T>;
+}
+macro_rules! impl_write_as_sexp_for_iterable {
+    ($iterable:ty, $item:ident $(, const $n:ident: $n_type:ty)?) => {
+        impl<$item $(, const $n: $n_type)?> WriteAsSExp<$item> for $iterable
+        where
+            $item: WriteAsIon,
+        {
+            fn as_sexp(self) -> SExpTypeHint<Self, T> {
+                SExpTypeHint::new(self)
+            }
+        }
+    };
+}
+
+impl_write_as_sexp_for_iterable!(Vec<T>, T);
+impl_write_as_sexp_for_iterable!(&[T], T);
+impl_write_as_sexp_for_iterable!([T; N], T, const N: usize);
+
+pub struct SExpTypeHint<S, T> {
```

Comment:
What is this for?

Reply:
Oops, this needs a doc comment. It wraps types that would be serialized as a `list` by default in a type that is always serialized as a `sexp`.

```rust
writer
  .write([1, 2, 3])?
  .write([1, 2, 3].as_sexp())?
```
would produce
```ion_1_1
[1, 2, 3]
(1 2 3)
```